Plutarch/Api/V1/Value.hs

```diff
@@ -161,6 +184,105 @@ instance
   where
   inv a = punsafeCoerce $ PlutusTx.inv (punsafeCoerce a :: Term s (PValue 'Sorted 'NoGuarantees))
 
+-- | Check whether a binary relation holds over 2 sorted 'PValue's.
+pcheckBinRel ::
+  Term
+    s
+    ( (PAsData PInteger :--> PAsData PInteger :--> PBool)
+        :--> PValue 'Sorted any0
+        :--> PValue 'Sorted any1
+        :--> PBool
+    )
+pcheckBinRel = phoistAcyclic $
+  plam $ \f m1 m2 ->
+    let inner = pfix #$ plam $ \self l1 l2 ->
+          pelimList
+            ( \x xs ->
+                pelimList
+                  ( \y ys -> unTermCont $ do
+                      v1dat <- tcont . plet $ psndBuiltin # x
+                      v2dat <- tcont . plet $ psndBuiltin # y
+                      k1 <- tcont . plet $ pfromData $ pfstBuiltin # x
+                      k2 <- tcont . plet $ pfromData $ pfstBuiltin # y
+                      pure $
+                        pif
+                          (k1 #< k2)
+                          (pinnerCheckBinRel # f # pfromData v1dat # AssocMap.pempty #&& self # xs # l2)
+                          $ pif
+                            (k1 #== k2)
+                            ( pinnerCheckBinRel # f # pfromData v1dat # pfromData v2dat #&& self
+                                # xs
+                                # ys
+                            )
+                            $ pinnerCheckBinRel # f # AssocMap.pempty # pfromData v2dat #&& self
+                              # l1
+                              # ys
+                  )
+                  ( pinnerCheckBinRel # f # pfromData (psndBuiltin # x) # AssocMap.pempty
+                      #&& pall
+                        # plam
+                          ( \p ->
+                              let v1dat = psndBuiltin # p
+                               in pinnerCheckBinRel # f # pfromData v1dat # AssocMap.pempty
+                          )
+                        # xs
+                  )
+                  l2
+            )
+            ( pall
+                # plam
+                  ( \p ->
+                      let v1dat = psndBuiltin # p
+                       in pinnerCheckBinRel # f # AssocMap.pempty # pfromData v1dat
+                  )
+                # l2
+            )
+            l1
+     in inner # pto (pto $ pto m1) # pto (pto $ pto m2)
+  where
+    dat0 :: ClosedTerm (PAsData PInteger)
+    dat0 = pconstantData 0
+    pinnerCheckBinRel ::
+      ClosedTerm
+        ( (PAsData PInteger :--> PAsData PInteger :--> PBool)
+            :--> ( PMap 'Sorted PTokenName PInteger
+                    :--> (PMap 'Sorted PTokenName PInteger :--> PBool)
+                 )
+        )
+    pinnerCheckBinRel = phoistAcyclic $
```

Comment:
There is a lot of common logic shared between this and the main one. I think you can formulate this as something that works on `PMap a b` where `POrd a`, given some "default value".

Reply:
it's intentionally written like this to optimize for CPU/Mem. But now that I think about it, there might be a minimal trade off in CPU/Mem for a decent script size saving here - I implemented that.

Reply:
This reasoning doesn't make sense to me? Why would you save CPU by duplicating the logic?
If you're thinking that you'll somehow have more applications because you use a more abstraction version of the function,
that is possible (albeit I'm not sure it's true), but in that case, you still don't want to duplicate the logic.
You want to instantiate and monomorphise the logic at several types.

Reply:
That doesn't address my comment.